### DISCUSSION TITLE
Unable to link share point pages/aspx to "Hero" template programmatically (using dot net 6.0).

### DISCUSSION THREAD

**Discussion by vasudev-opteamix** [pnp/pnpcore#1321] [Q&A]
Hello @jansenbe ,

We are trying to programmatically  add content to SharePoint from old CMS . There are lot of pages that we need to migrate and we came across this PnpCore Sdk. We are unable to find a proper documentation for this. In future we want to add more sharepoint webparts programmatically.

We were able to "Hero" web part programmatically but we are unable to link any articles/aspx pages programmatically. We are using this using Dot Net 6.0. Could you please help us with this ?? 

Below is the code used for object creation of "Hero" template and binding the same to my custom page.

```
templatePage.AddSection(CanvasSectionTemplate.OneColumnFullWidth, 1);
var availableComponents = await templatePage.AvailablePageComponentsAsync();
var heroWebPartComponent = availableComponents.FirstOrDefault(p => p.Id == templatePage.DefaultWebPartToWebPartId(DefaultWebPart.Hero));
templatePage.AddControl(templatePage.NewWebPart(heroWebPartComponent));
```

**Comment by jansenbe** [comment]
@vasudev-opteamix : here's the documentation for the Pages API support in PnP Core SDK: https://pnp.github.io/pnpcore/using-the-sdk/pages-intro.html. As you'll need to create the correct web part properties JSON it's recommended to first understand what's needed by configuring the needed (custom) web parts in the work bench and then inspect the JSON (see https://pnp.github.io/pnpcore/using-the-sdk/pages-webparts.html#generic-instructions-that-apply-for-all-web-parts to learn more)

**Reply by vasudev-opteamix** [reply]
Hi @jansenbe ,   we are able to add image webpart using above approach (find attached code). We are not able to do similar thing for the "Hero" webpart. 

we are unable to link any articles/aspx pages programmatically.

 Can you please help us on this ??  It is good, if you provide us some code example.  

```
             ```
  var newpage = await context.Web.NewPageAsync();
                // adding sections to the page
                newpage.AddSection(CanvasSectionTemplate.OneColumn, 1);
 
                // configure an image web part for a given site relative url
                var image2 = await newpage.GetImageWebPartAsync("/sites/UmamiFoodMagazine2/siteassets/cad.jpg", new 
               PageImageOptions
                {
                    Alignment = PageImageAlignment.Left,
                    Width = 200,
                    Height = 200
                });
 
                // add the web part to the first column of the first section
                newpage.AddControl(image2, newpage.Sections[0].Columns[0]);
 
                // Save the page
                await newpage.SaveAsync("MyPage-MyImage.aspx");. 
```

Thank you.
```

**Reply by jansenbe** [reply]
@vasudev-opteamix : for the image web part we do have "typed" support, but we don't have that for the other ones. So you would need to craft the web part properties json and pass that in as explained in the "generic web part instructions" over here: https://pnp.github.io/pnpcore/using-the-sdk/pages-webparts.html#generic-instructions-that-apply-for-all-web-parts.

Depending on the web part creating a valid properties json can be complex as you'll need to insert the right file id's (guids) and/or paths. Best is to first configure what you need via the workbench, copy the properties json and then adapt that to make it work for your data
